skimage/morphology/_extrema_cy.pyx

```diff
@@ -0,0 +1,190 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+cimport numpy as cnp
+from libc.stdlib cimport malloc, realloc, free
+
+
+ctypedef fused dtype_t:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.double_t
+
+
+ctypedef Py_ssize_t QueueItem
+
+
+cdef:
+    struct Queue:
```

Comment:
Any chance we could re-use a data structure already present somewhere else in order to limit the complexity of the code?

Reply:
Oh I see that you've mentioned the Queue of the watershed. A common codebase would be great if it makes sense.

Reply:
I agree, a common codebase would be ideal (see this comment as well https://github.com/scikit-image/scikit-image/issues/2876#issuecomment-385241507). However as already mentioned there are some problematic differences between the two queue implementations and its requirements which I will try to summarize below:

Heap (general_heap.pxi) is a priority queue used by watershed, Queue is used in my code:
- Multiple iterations:  Queue allows multiple iterations via `q_restore` and preserves a consumed (popped) item internally until cleared explicitly. Heap may overwrite consumed data. 
- Order: Queue uses insertion order while Heap sorts (using the array of pointers) by the return value of a function `smaller`.
- A consequence the internal data representation is different as well: Queue uses only one array to store its items while Heap uses an additional array to store pointers to its items. 

@emmanuelle You seem to be the original contributor of Heap so please point out any misunderstandings on my part.

I don't see much overlap here concerning the internal machinery of these "basic" containers. But it may be possible to implement a data container that fulfills the requirements of both Heap and Queue. But would that be wise? I think my code would be slower due to overhead introduced by the sorting machinery and watershed would have to suffer the larger memory footprint due to internal preservation of popped items.

Having said that, if you still think it might be worth it I'll try to make a quick draft of a queue that can be used as a replacement for both. That might be more useful to reach a more informed decision.

Reply:
No no, a FIFO Queue and a Heap are two different data structures, even if the latter is used to implement Priority Queues. I don't think there is much code to be shared between them and it makes sense to keep them separate.